### PR TITLE
Updates to campaign resource for new index listing.

### LINKED ISSUE #4541: Replace prior campaign endpoint code to use new Campaign class
Replace both _index_ and _retrieve_ campaign endpoint code with new approach.

### PR DESCRIPTION
**Pull request by weerd**
### Fixes #4541

This PR wraps up shifting over both the `/campaigns` and `/campaigns/:id` endpoints and standardized JSON response.

The _index_ listing of campaigns include all pagination data in same format as that of the reportbacks/reportback-items index views. It also provides the following parameters to filter by:
- ids (string; csv)
- count (default is 25)
- staff_pick (boolean)
- mobile_app (boolean)
- term_ids (string; csv)
- random (boolean)
- page (integer)

Example request:

```
/api/v1/campaigns.json?&count=2&page=2
```

Example response:

```
{
  pagination: {
    total: 15,
    per_page: 2,
    current_page: 1,
    total_pages: 8,
    prev_uri: null,
    next_uri: "http://dev.dosomething.org:8888/api/v1/campaigns.json?&count=2&page=2"
  },
  retrieved_count: 2,
  data: [
    {
      id: "1261",
      title: "1 in 3 of Us",
      tagline: "This is a sandbox campaign.",
      created_at: "1404323239",
      updated_at: "1430407941",
      time_commitment: 2,
      status: "active",
      cover_image: {
        default: {
          uri: "http://dev.dosomething.org:8888/sites/default/files/styles/300x300/public/images/square_image_JK8675309.jpg",
          sizes: {
            landscape: {
              uri: "http://dev.dosomething.org:8888/sites/default/files/styles/1440x810/public/images/landscape_image_JK8675309.jpg"
            },
            square: {
              uri: "http://dev.dosomething.org:8888/sites/default/files/styles/300x300/public/images/square_image_JK8675309.jpg"
            }
          },
          type: "image",
          dark_background: true
        }
      },
      staff_pick: true,
      facts: {
        problem: "In the United Kingdom, people drink tea.",
        solution: "Experts agree that drinking at 3 cups of coffee per day prevents all sleepiness-related disasters.",
        sources: [
          {
            formatted: "<p>Clark, Barry. "A Treatise on The Motherland", British Royal Archives, 2009.</p> "
          },
          {
            formatted: "<p>Lorch, Julie. "Coffee Sure Is Great", Royal Academy of Coffee (2010)</p> "
          }
        ]
      },
      solutions: {
        copy: {
          raw: "Be American! Drink coffee to stay awake! DoSomething.org recommends 3 or more cups per day.",
          formatted: "<p>Be American! Drink coffee to stay awake! DoSomething.org recommends 3 or more cups per day.</p> "
        }
      },
      causes: {
        secondary: [
          {
            id: "19",
            name: "mental health"
          },
          {
            id: "5",
            name: "physical health"
          }
        ]
      },
      action_types: {
        primary: {
          id: "3",
          name: "face to face"
        },
        secondary: [
          {
            id: "3",
            name: "face to face"
          },
          {
            id: "8",
            name: "improve a space"
          },
          {
            id: "9",
            name: "make something"
          },
          {
            id: "18",
            name: "share something"
          }
        ]
      },
      issue: {
        id: "490",
        name: "coffee"
      },
      tags: [
        {
          id: "491",
          name: "coffee"
        },
        {
          id: "492",
          name: "maxwell watson"
        },
        {
          id: "493",
          name: "birch"
        },
        {
          id: "494",
          name: "beans"
        },
        {
          id: "495",
          name: "aeropress"
        },
        {
          id: "496",
          name: "chemex"
        }
      ]
    },
    {
      id: "1617",
      title: "Another Test Campaign",
      tagline: "testing the pre-signup/signup script",
      created_at: "1422907938",
      updated_at: "1422911103",
      time_commitment: 0,
      status: "active",
      cover_image: {
        default: {
          uri: "http://dev.dosomething.org:8888/sites/default/files/styles/300x300/public/square_image_GH4yP13s.jpg",
          sizes: {
            landscape: {
              uri: "http://dev.dosomething.org:8888/sites/default/files/styles/1440x810/public/landscape_image_GH4yP13s.jpg"
            },
            square: {
              uri: "http://dev.dosomething.org:8888/sites/default/files/styles/300x300/public/square_image_GH4yP13s.jpg"
            }
          },
          type: "image",
          dark_background: false
        }
      },
      staff_pick: true
    }
  ]
}
```

@aaronschachter @jonuy 

CC: @angaither @DFurnes

### CODE REVIEW

File: lib/modules/dosomething/dosomething_campaign/dosomething_campaign.module [hunk 1]
```diff
@@ -362,65 +380,76 @@ function dosomething_campaign_reportback_confirmation_page_access($node) {
...
 
+
 /**
  * Returns array of published campaign nid and titles by campaign type $type.
  */
-function dosomething_campaign_get_campaigns_query_result($params = array()) {
+function dosomething_campaign_get_campaigns_query_result($params = array(), $count = 25, $start = 0) {
   $query = dosomething_campaign_get_campaigns_query($params);
 
+  if ($count && $count !== 'all') {
```
Comment (aaronschachter) [inline lib/modules/dosomething/dosomething_campaign/dosomething_campaign.module:390]: Is this correct? Should be `!=` ?

File: lib/modules/dosomething/dosomething_api/resources/campaign_resource.inc [hunk 2]
```diff
@@ -1,65 +1,105 @@
 <?php
 
 function _campaign_resource_definition() {
-  $campaign_resource = array();
-  $campaign_resource['campaigns'] = array(
-    'operations' => array(
+  $campaign_resource = [];
```
Comment (angaither) [inline lib/modules/dosomething/dosomething_api/resources/campaign_resource.inc:4]: this is legal now!? YAY

File: lib/modules/dosomething/dosomething_campaign/dosomething_campaign.module (hunk 1 above)
Comment (weerd) [inline lib/modules/dosomething/dosomething_campaign/dosomething_campaign.module:390]: checks that's it's of correct type too, so a string :)

File: lib/modules/dosomething/dosomething_api/resources/campaign_resource.inc (hunk 2 above)
Comment (weerd) [inline lib/modules/dosomething/dosomething_api/resources/campaign_resource.inc:4]: Yeeeeep! Thanks to @sergii-tkachenko :dancers:  Officially on PHP 5.5 or whatevs on all AWS thingies <--- technical term

**Comment by aaronschachter** [comment]
Looks ok to me except for question..

File: lib/modules/dosomething/dosomething_campaign/dosomething_campaign.module (hunk 1 above)
Comment (aaronschachter) [inline lib/modules/dosomething/dosomething_campaign/dosomething_campaign.module:390]: you and your fancy php :necktie:

Comment (weerd) [inline lib/modules/dosomething/dosomething_campaign/dosomething_campaign.module:390]: hehe :dancer: olé!

**Comment by aaronschachter** [comment]
:+1:

**Pull request merged**